File: lib/core/chess/game_state.dart
import 'package:dartchess/dartchess.dart' as dc;

enum GameStatus { playing, checkmate, stalemate, draw }

/// Stateful game wrapper around dartchess. dartchess validates moves and
/// reports check/mate, but does NOT auto-detect repetition or the 50/75-move
/// rules — that is this layer's job (silent-correctness trap).
class GameState {
  GameState() : this.fromFen(dc.Setup.standard.fen);

  GameState.fromFen(String fen)
    : _position = dc.Chess.fromSetup(dc.Setup.parseFen(fen)) {
    _fenHistory.add(_repetitionKey);
  }

  dc.Position _position;
  final List<String> _fenHistory = <String>[];
  final List<dc.Move> _moves = <dc.Move>[];

  dc.Position get position => _position;
  String get fen => _position.fen;
  String get sideToMove => _position.turn.name;
  bool get isGameOver => status != GameStatus.playing;
  List<String> get pgnMoves => _moves.map((m) => m.uci).toList();

  /// Every position reached in this game (oldest first, the current position
  /// last). Feeds the az119 encoder for the RL AI.
  List<String> get fenHistory => List.unmodifiable(_fenHistory);

  List<dc.Move> get legalMoves {
    final result = <dc.Move>[];
    for (final entry in _position.legalMoves.entries) {
      final from = entry.key;
      final piece = _position.board.pieceAt(from);
      for (final to in entry.value.squares) {
        final promotes =
            piece != null &&
            piece.role == dc.Role.pawn &&
            to.rank ==
                (piece.color == dc.Side.white ? dc.Rank.eighth : dc.Rank.first);
        if (promotes) {
          for (final role in _promotionRoles) {
            result.add(dc.NormalMove(from: from, to: to, promotion: role));
          }
        } else {
          result.add(dc.NormalMove(from: from, to: to));
        }
      }
    }
    return result;
  }

  static const _promotionRoles = <dc.Role>[
    dc.Role.queen,
    dc.Role.rook,
    dc.Role.bishop,
    dc.Role.knight,
  ];

  GameStatus get status {
    if (_position.isCheckmate) return GameStatus.checkmate;
    if (_position.isStalemate) return GameStatus.stalemate;
    if (isThreefoldRepetition || isFiftyMoveRule) return GameStatus.draw;
    return GameStatus.playing;
  }

  bool get isThreefoldRepetition {
    // _fenHistory holds every position reached (including the current one and
    // the start position); >= 3 occurrences of the current position is a draw.
    final counts = <String, int>{};
    for (final f in _fenHistory) {
      counts[f] = (counts[f] ?? 0) + 1;
    }
    return (counts[_repetitionKey] ?? 0) >= 3;
  }

  /// Position identity for the repetition rule: FEN without the halfmove
  /// clock and fullmove counter — those change on every ply, so the same
  /// board position reached twice would otherwise look like two positions.
  String get _repetitionKey => _position.fen.split(' ').take(4).join(' ');

  /// FIDE fifty-move rule: a draw is claimable after 100 plies (half-moves)
  /// without a capture or pawn move. dartchess tracks the clock as `halfmoves`.
  bool get isFiftyMoveRule => _position.halfmoves >= 100;

  /// Why the game is drawn, or `null` if no draw condition is currently met.
  /// Computed so it can never go stale after undo.
  String? get drawReason {
    if (isThreefoldRepetition) return 'repetition';
    if (isFiftyMoveRule) return '50-move rule';
    return null;
  }

  bool makeMove(String uci) {
    final m = uci.length < 4 ? null : dc.Move.parse(uci);
    if (m == null || !_position.isLegal(m)) return false;
    // Reject a bare pawn push to the promotion rank: dartchess accepts it, but
    // play() then leaves an illegal pawn on the last rank. GameState is the
    // trusted rules layer (future multiplayer), so enforcement lives here.
    if (m is dc.NormalMove && m.promotion == null) {
      final piece = _position.board.pieceAt(m.from);
      final promotes =
          piece != null &&
          piece.role == dc.Role.pawn &&
          m.to.rank ==
              (piece.color == dc.Side.white ? dc.Rank.eighth : dc.Rank.first);
      if (promotes) return false;
    }
    _position = _position.play(m);
    _moves.add(m);
    _fenHistory.add(_repetitionKey);
    return true;
  }

  void undoLast() {
    if (_moves.isEmpty) return;
    _moves.removeLast();
    _fenHistory.removeLast();
    _position = _replay(_fenHistory.first, _moves);
  }

  static dc.Position _replay(String startFen, List<dc.Move> moves) {
    dc.Position pos = dc.Chess.fromSetup(dc.Setup.parseFen(startFen));
    for (final m in moves) {
      pos = pos.play(m);
    }
    return pos;
  }
}

File: packages/core/src/chess/game-state.ts
import { Chess, type Move } from "chess.js";

export enum GameStatus {
  Playing = "playing",
  Checkmate = "checkmate",
  Stalemate = "stalemate",
  Draw = "draw",
}

/**
 * Stateful game wrapper around chess.js.
 * Mirrors the Dart GameState: tracks position, move history, FEN history
 * for repetition detection, and enforces promotion rules.
 */
export class GameState {
  private _chess: Chess;
  private _fenHistory: string[] = [];
  private _moves: Move[] = [];

  constructor(fen?: string) {
    this._chess = new Chess(fen);
    this._fenHistory.push(this.repetitionKey);
  }

  private get repetitionKey(): string {
    return this._chess.fen().split(" ").slice(0, 4).join(" ");
  }

  get fen(): string {
    return this._chess.fen();
  }

  get sideToMove(): "white" | "black" {
    return this._chess.turn() === "w" ? "white" : "black";
  }

  get isGameOver(): boolean {
    return this.status !== GameStatus.Playing;
  }

  get pgnMoves(): string[] {
    return this._moves.map((m) => m.from + m.to + (m.promotion ?? ""));
  }

  get fenHistory(): string[] {
    return [...this._fenHistory];
  }

  get legalMoves(): Move[] {
    return this._chess.moves({ verbose: true });
  }

  get status(): GameStatus {
    if (this._chess.isCheckmate()) return GameStatus.Checkmate;
    if (this._chess.isStalemate()) return GameStatus.Stalemate;
    if (this.isThreefoldRepetition || this.isFiftyMoveRule)
      return GameStatus.Draw;
    return GameStatus.Playing;
  }

  get isThreefoldRepetition(): boolean {
    const counts = new Map<string, number>();
    for (const f of this._fenHistory) {
      counts.set(f, (counts.get(f) ?? 0) + 1);
    }
    return (counts.get(this.repetitionKey) ?? 0) >= 3;
  }

  get isFiftyMoveRule(): boolean {
    const history = this._chess.history({ verbose: true });
    if (history.length === 0) return false;
    const last = history[history.length - 1];
    // chess.js tracks halfmoves internally; we check via FEN
    const halfmoves = parseInt(this.fen.split(" ")[4], 10);
    return halfmoves >= 100;
  }

  get drawReason(): string | null {
    if (this.isThreefoldRepetition) return "repetition";
    if (this.isFiftyMoveRule) return "50-move rule";
    return null;
  }

  makeMove(uci: string): boolean {
    if (uci.length < 4) return false;
    const from = uci.slice(0, 2);
    const to = uci.slice(2, 4);
    const promotion = uci.length === 5 ? uci[4] : undefined;

    // Reject bare pawn push to promotion rank without promotion piece
    const piece = this._chess.get(from as Square);
    if (
      piece?.type === "p" &&
      !promotion &&
      ((piece.color === "w" && to[1] === "8") ||
        (piece.color === "b" && to[1] === "1"))
    ) {
      return false;
    }

    try {
      const move = this._chess.move({ from, to, promotion });
      if (!move) return false;
      this._moves.push(move);
      this._fenHistory.push(this.repetitionKey);
      return true;
    } catch {
      return false;
    }
  }

  undoLast(): void {
    if (this._moves.length === 0) return;
    this._moves.pop();
    this._fenHistory.pop();
    // Replay from scratch
    const startFen = this._fenHistory[0];
    this._chess = new Chess(startFen);
    for (const m of this._moves) {
      this._chess.move(m);
    }
  }

  clone(): GameState {
    const gs = new GameState(this.fen);
    return gs;
  }
}

type Square =
  | "a1" | "a2" | "a3" | "a4" | "a5" | "a6" | "a7" | "a8"
  | "b1" | "b2" | "b3" | "b4" | "b5" | "b6" | "b7" | "b8"
  | "c1" | "c2" | "c3" | "c4" | "c5" | "c6" | "c7" | "c8"
  | "d1" | "d2" | "d3" | "d4" | "d5" | "d6" | "d7" | "d8"
  | "e1" | "e2" | "e3" | "e4" | "e5" | "e6" | "e7" | "e8"
  | "f1" | "f2" | "f3" | "f4" | "f5" | "f6" | "f7" | "f8"
  | "g1" | "g2" | "g3" | "g4" | "g5" | "g6" | "g7" | "g8"
  | "h1" | "h2" | "h3" | "h4" | "h5" | "h6" | "h7" | "h8";
